### PR TITLE
Fix Documentation in FSDP and DeepSpeed Concept Guide

### PR DESCRIPTION
**Pull request by fabianlim**
# What does this PR do?

Address @stas00's comments that was omitted when we merged  #2674 

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@stas00 @muellerzr

### CODE REVIEW

**Comment by fabianlim** [comment]
@muellerzr the [documentation build failed](https://github.com/huggingface/accelerate/actions/runs/8889182453/job/24419633253?pr=2725). I think its because now `deepspeed` is an [optional dep](https://github.com/huggingface/accelerate/blob/4d3d4bc88f557692eff1dc2fd8c5f80153045e4e/setup.py#L39) now. It needs to be installed in the github workflow somehow.

**Comment by muellerzr** [comment]
To get around this instead of [`utils.DummyOptim`] just do [`utils.deepspeed.DummyOptim`]. There's no need for deepspeed to be installed

File: docs/source/concept_guides/fsdp_and_deepspeed.md [hunk 1]
```diff
@@ -81,7 +94,7 @@ FSDP only allows *all-or-nothing* offload (i.e., either offload parameters, grad
 ### Prefetching
 
 FSDP allows two prefetching configurations `--fsdp_forward_prefetch` and `--fsdp_backward_prefetch` to improve overlap of comms / computation at a cost of extra memory, see [FSDP documentation](https://pytorch.org/docs/stable/fsdp.html). 
-For DeepSpeed, the prefetching is always on, and only certain hyperparams like `stage3_prefetch_bucket_size` [can be configured for Zero3](https://www.deepspeed.ai/docs/config-json/#parameter-offloading); 🤗 `accelerate` will set these hyperparams automatically.
+For DeepSpeed, the prefetching will be turned on when needed, and it turns on depending on certain hyperparams like `stage3_param_persistence_threshold, stage3_max_reuse_distance`, etc, [that can be configured for Zero3](https://www.deepspeed.ai/docs/config-json/#parameter-offloading); 🤗 `accelerate` will set these hyperparams automatically.
```
Comment (stas00) [inline docs/source/concept_guides/fsdp_and_deepspeed.md:97]:
```suggestion
For DeepSpeed, the prefetching will be turned on when needed, and it turns on depending on certain hyper-params like `stage3_param_persistence_threshold`, `stage3_max_reuse_distance`, etc, [that can be configured for Zero3](https://www.deepspeed.ai/docs/config-json/#parameter-offloading); 🤗 `accelerate` may set these hyper-params automatically if you don't set those explicitly in the deepspeed config file.
```

File: docs/source/concept_guides/fsdp_and_deepspeed.md [hunk 2]
```diff
@@ -133,7 +147,7 @@ Deepspeed requires explicit `--gradient_accumulation_steps` and `--gradient_clip
...
+To discuss the how data precision is handled in both FSDP and Deepspeed, it is instructive to first give an overview of how model parameters are handled in these frameworks. Before the model / optimizer parameters are distributed across GPUs, parameter preparation is involved to first "flatten" them to  one-dimensional [`torch.Tensor`](https://pytorch.org/docs/stable/tensors.html#torch-tensor). The implementation of FSDP / DeepSpeed varies in the respect of the `dtype` in which these "flattened" parameters are stored, and there are ramifications with regards to how [`torch.Optimizer`](https://pytorch.org/docs/stable/optim.html#module-torch.optim) allocate their `dtypes`. The table below outlines the processes for both frameworks; the "Local" column indicates the process occurring at a per-gpu level, therefore any memory overheads by upcasting should be understood to be amortized by the number of gpus used.
```
Comment (stas00) [inline docs/source/concept_guides/fsdp_and_deepspeed.md:150]:
```suggestion
To discuss the how data precision is handled in both FSDP and Deepspeed, it is instructive to first give an overview of how model parameters are handled in these frameworks. Before the model / optimizer parameters are distributed across GPUs, parameter preparation is involved to first "flatten" them to  one-dimensional [`torch.Tensor`](https://pytorch.org/docs/stable/tensors.html#torch-tensor). The implementation of FSDP / DeepSpeed varies in the respect of the `dtype` in which these "flattened" parameters are stored, and there are ramifications with regards to how [`torch.Optimizer`](https://pytorch.org/docs/stable/optim.html#module-torch.optim) allocate their `dtype`s. The table below outlines the processes for both frameworks; the "Local" column indicates the process occurring at a per-gpu level, therefore any memory overheads by upcasting should be understood to be amortized by the number of gpus used.
```

**Review by stas00 — APPROVED** [review]
Thank you for the follow up PR, @fabianlim!

I added a few small suggestions if they resonate.

**Comment by fabianlim** [comment]
@stas00 thank you for your suggestions I have integrated them.

> To get around this instead of [utils.DummyOptim] just do [utils.deepspeed.DummyOptim]. There's no need for deepspeed to be installed

@muellerzr I tried this but it seemed I need to fix the `find_object_in_package` function in `doc_builder`. I have made a [PR](https://github.com/huggingface/doc-builder/pull/501). 

I have also added the links for the FSDP and DS docs https://github.com/huggingface/accelerate/pull/2725/commits/3b1948ff7921a6edb18aff28e9096d19245cab15. Hope they are added in appropriate places. If not please let me know.

Im still on the fence whether to put back the raise

**Comment by muellerzr** [comment]
I'm seeing this failure on another PR, likely not from you at all and instead stems from existing docs

**Comment by muellerzr** [comment]
@fabianlim can you rebase from `main`? We just merged something that should fix it!

**Comment by fabianlim** [comment]
@muellerzr just rebased!

**Review by muellerzr — APPROVED** [review]
Thanks a bunch for all your hard work!

**Pull request merged**